Userland/Libraries/LibWeb/CSS/Parser/Parser.cpp

```diff
@@ -171,392 +177,470 @@ NonnullRefPtr<CSSStyleSheet> Parser::parse_as_stylesheet(TokenStream<T>& tokens)
 
-                if (next.is(Token::Type::Delim) && next.token().delim() == "|") {
-                    relation = Selector::ComplexSelector::Relation::Column;
-                    tokens.next_token();
-                }
-            }
+        bool is_pseudo = false;
+        if (tokens.peek_token().is(Token::Type::Colon)) {
+            is_pseudo = true;
+            tokens.next_token();
+            if (peek_token_ends_selector())
+                return SelectorParsingResult::SyntaxError;
         }
 
-        tokens.skip_whitespace();
+        if (is_pseudo) {
+            Selector::SimpleSelector simple_selector {
+                .type = Selector::SimpleSelector::Type::PseudoElement
+            };
 
-        Vector<Selector::SimpleSelector> simple_selectors;
+            auto& name_token = tokens.next_token();
+            if (!name_token.is(Token::Type::Ident)) {
+                dbgln_if(CSS_PARSER_DEBUG, "Expected an ident for pseudo-element, got: '{}'", name_token.to_debug_string());
+                return SelectorParsingResult::SyntaxError;
+            }
 
-        for (;;) {
-            auto current_value = tokens.peek_token();
-            if (current_value.is(Token::Type::EndOfFile) || current_value.is(Token::Type::Whitespace))
-                break;
+            auto pseudo_name = name_token.token().ident();
 
-            auto component = parse_simple_selector();
-            if (!component.has_value())
-                break;
+            if (pseudo_name.equals_ignoring_case("before")) {
+                simple_selector.pseudo_element = Selector::SimpleSelector::PseudoElement::Before;
+            } else if (pseudo_name.equals_ignoring_case("after")) {
+                simple_selector.pseudo_element = Selector::SimpleSelector::PseudoElement::After;
+            } else if (pseudo_name.equals_ignoring_case("first-line")) {
+                simple_selector.pseudo_element = Selector::SimpleSelector::PseudoElement::FirstLine;
+            } else if (pseudo_name.equals_ignoring_case("first-letter")) {
+                simple_selector.pseudo_element = Selector::SimpleSelector::PseudoElement::FirstLetter;
+            } else {
+                dbgln_if(CSS_PARSER_DEBUG, "Unrecognized pseudo-element: '{}'", pseudo_name);
+                return SelectorParsingResult::SyntaxError;
+            }
 
-            simple_selectors.append(component.value());
+            return simple_selector;
         }
 
-        if (simple_selectors.is_empty())
-            return {};
+        if (peek_token_ends_selector())
+            return SelectorParsingResult::SyntaxError;
+
+        auto& pseudo_class_token = tokens.next_token();
+        Selector::SimpleSelector simple_selector {
+            .type = Selector::SimpleSelector::Type::PseudoClass
+        };
+
+        if (pseudo_class_token.is(Token::Type::Ident)) {
+            auto pseudo_name = pseudo_class_token.token().ident();
+            if (pseudo_name.equals_ignoring_case("link")) {
```

Comment:
Is there a logic to the order of these? Just curious...

Reply:
Well spotted! That's copied from the previous selector parsing code, which I think is copied from the deprecated parser's code. But yeah, the order makes no sense! I'll switch it to alphabetical. (Maybe there's a "most common first" order we could do, but we have no data for that, and it might make more sense to use a hash table then anyway.)

Reply:
Linus'll love you for that ;)

Reply:
A couple of things were in there twice, so it's 3 benefits in one! (Correctness, comfiness, and happy Linus)